platform-sdk/docs/proposals/001-merkle-state-api/001-merkle-state-api.md

```diff
@@ -0,0 +1,311 @@
+   * Gets whether the value of this {@link ReadableSingletonState} has been read.
+   *
+   * @return true if {@link #get()} has been called on this instance
+   */
+  boolean isRead();
+}
+/**
+ * Provides mutable access to singleton state.
+ *
+ * @param <T> The type of the state
+ */
+public interface WritableSingletonState<T> extends ReadableSingletonState<T> {
+  /**
+   * Sets the given value on this state.
+   *
+   * @param value The value. May be null.
+   */
+  void put(@Nullable T value);
+
+  /**
+   * Gets whether the {@link #put(Object)} method has been called on this instance.
+   *
+   * @return True if the {@link #put(Object)} method has been called
+   */
+  boolean isModified();
+}
+```
+
+## Requirements
+
+- `PlatformState` should be accessible via the State API 
+- there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hashgraph state.
+  This module should have a minimal set of dependencies. The Block Node should not have a compile-time dependency
+  on any other modules but this one to interact with the state.
+
+## Changes
+
+### Public API
+
+State API classes and interfaces will migrate to a designated module. The classes themselves will not change. `PlatformState` class will become a part of the public API.
+
+Usage example:
+```java
+private PlatformState findPlatformState(HederaState state) {
+  final ReadableStates states = state.getReadableStates(PlatformState.NAME);   
+  final ReadableSingletonState<PlatformState> platformState = states.getSingleton(PlatformState.PLATFORM_STATE_KEY);
+  return platformState.get();
+}
+```
+
+## Components and Architecture
+
+[### Platform State](#platform-state)
+
+Currently `PlatformState` is a special case node of the merkle tree. That is, it doesn't conform to State API. 
+As a part of the simplification effort **it should be refactored to a singleton state**. 
+As a part of this change, `PlatformState` should not be a `MerkleNode` anymore.
+
+Protobuf definition:
+
+```
+/**
+ * The current state of platform consensus.<br/>
+ * This message stores the current consensus data for the platform
+ * in network state.
+ *
+ * The platform state SHALL represent the latest round's consensus.<br/>
+ * This data SHALL be used to ensure consistency and provide critical data for
+ * restart and reconnect.
+ */
+message PlatformState {
+    /**
+     * A consensus round.<br/>
+     * The round represented by this state.
+     * <p>
+     * This message SHALL represent the network state after handling of all
+     * transactions reaching consensus in all previous rounds.<br/>
+     * The first state (genesis state) SHALL have a round of 0.<br/>
+     * The first round to handle transactions SHALL be round `1`.
+     */
+    uint64 round = 1;
+
+    /**
+     * A consensus timestamp for this state.
+     * <p>
+     * This SHALL be the consensus timestamp of the first transaction in
+     * the current round.
```

Comment:
Not all rounds will have transactions, so I don't think this is right. There could even be rounds with no events.

The javadoc in the `ConsensusRound` class says:

```
The timestamp of the end of a round. Is equal to the consensus timestamp of the last transaction in the last event in the round, or the consensus timestamp of the last event in the round if there are no transactions in the last event. The consensus timestamp of a round is guaranteed to be strictly greater than the consensus timestamp of the previous round.
```
but this does not cover empty rounds. We need to figure out how to handle that case. I'm not sure what the answer should be.

Reply:
I removed this field as well.

Reply:
- if there are transactions, the timestamp is equal to the timestamp of the last transaction
- if there are no transactions, the timestamp is equal to the timestamp of the last event
- if there are no events, the timestamp is equal to the timestamp of the previous round plus a small constant (1ns, 100ns, whatever. Should be a setting that services configures.)

Reply:
@cody-littley this is great. I think this should be specified in the javadoc so external consumers know exactly how to verify these situations.

Reply:
Done. I updated javadoc for `ConsensusSnapshot.consensus_timestamp`